### DISCUSSION TITLE
Link actuation with a point/frame of a mesh

### DISCUSSION THREAD

**Discussion by bmauze** [SofaDefrost/SoftRobots#157] [Help - Programming with SoftRobots]
Hello everyone,

I am trying to create a soft delta using a SlidingActuator as actuation system. The code i used is inspired from the tripod tutorial. However I do not get how to link the SlidingActuator to the frame associated to my actuation system on the elastic element.
I tried to use a RigidMapping on one side and a RigidRigidMapping on the othersize. But I have succeed to get something that was working. I also tried to use BarycentricMapping but the behavior is very strange . The constraint axis is not the right one (go toward Z while it should be the X one). 

So I have two questions: 
- If I have a elasticBody element whose some elements are rigidified and linked to a frame. Do you have a idea on how I could link this frame with a SlidingActuator element ?
- When I am using the Barycentric mapping, Why my structure is moving along the constrained wrong direction.

My code and the mesh can be found bellow1:
https://github.com/bmauze/SoftDelta.git

Kinds regards,
Benjamin Mauzé

**Comment by EulalieCoevoet** [comment]
Hello Benjamin, 

1- You can directly add the SlidingActuator in the frames' node. In your case:
```python
    frame = rigidifiedstruct.RigidParts  # your rigidified elements seem to be linked to the frames in this node
    # if you want to control those frames just add the actuators in the corresponding node    
    for i in range(3):
    	frame.addObject('SlidingActuator', name='Act'+str(i+1), template='Rigid3', 
    	                indices=i, direction=[0, 0, 1, 0, 0, 0], 
                        maxForce=forceActuator, minForce=-forceActuator, 
                        maxDispVariation=SetDepAct, maxPositiveDisp=depAct, maxNegativeDisp=depAct)
    # unfortunatly the component SlidingActuator needs the following component to fix the frames in the other directions
    frame.addObject('PartialFixedConstraint', template='Rigid3', indices=[0,1,2], fixedDirections=[1, 1, 0, 1, 1, 1])
```
Note that we typically use a mapping when the point that we want to control or actuate is not exactly a degree of freedom. 
Let's say you have an FEM mesh composed of tetrahedra. If the point you want to control/actuate is not a dof, but is inside a tetrahedron (right case in the image), you'll use a BarycentricMapping to map this point to your FEM mesh. 
![mapping](https://user-images.githubusercontent.com/12150715/174644245-6602d87a-074f-4c0d-ac12-7d76d143a387.png)
What this component does is computing the coordinates of this point inside the tetrahedron, and applying the displacements and forces from the point to the tetrahedron and vice-versa.  
Similarly with a rigid object you'll use a RigidRigidMapping.
![mapping (1)](https://user-images.githubusercontent.com/12150715/174646532-f1135f24-4070-454f-b627-619ce7d08963.png)

2- I don't know what kind of actuation you'll seek exactly, but if you want your elasticBody to be attached to an articulated chain, I suggest you to have a look at the ArticulatedSystemPlugin. We actually use it in the latest version of the Tripod tutorial, instead of the two components SlidingActuator and PartialFixedConstraint.

**Reply by bmauze** [reply]
Hello  Eulalie,

Thank you for this quick answer and all those details. If I understand well, in your response, the linking between the sliding actuators and the frames comes from the inheritance of the mechanical objects of _rigidifiedstruct.RigidParts_.
I changed a little bit my code to change manually the actuation system's position with the keyboard. A "problem" appeared. This problem reffers nearly to the same idea behind my previous question. The _MechanicalObject_ _freeCenter_ does not move even if the structure is moving. I thought that the link was made by the _MechanicalMatrixMapper_ but it is not the case (even if it seems to work well for an inverse mode in another program). I tried the _getLinkPath()_ function, but it linked all 4 positions and not only the third one. 
Do you have an idea to which direction I should look?

Thank you for the idea with the ArticulatedSystem Plugin. I am currently trying some stuff. Would the simulation work "better" using this new plugin? The actuation system that I want, is just a sliding actuator along a direction (here collinear to the X or Z axis)

**Reply by EulalieCoevoet** [reply]
No in that case you can stick to using the SlidingActuator :)

Well, if I understood correctly:
- currently all your rigidified elements are mapped to the `MechanicalObject` in the node `RigidParts`, which have the four frames
- what you want is to separate them into two groups, one with three of the frames and the other one with the fourth frame (in `FreeCenter`)?

If you look at the latest version of the Tripod tutorial we use the `SubsetMultiMapping` to do that. Here we map the four frames into four different nodes, the three `ActuatedArms` and the `FreeCenter`:

```python 
rigidParts.addObject('SubsetMultiMapping',
                              name="mapping",
                              input=[self.actuatedarms[0].ServoMotor.Articulation.ServoWheel.getLinkPath(),
                                     self.actuatedarms[1].ServoMotor.Articulation.ServoWheel.getLinkPath(),
                                     self.actuatedarms[2].ServoMotor.Articulation.ServoWheel.getLinkPath(),
                                     freeCenter.getLinkPath()],
                              output='@./', indexPairs=[0, 1, 1, 1, 2, 1, 3, 0])
```
Is this what you wanna do? If yes, may I ask why having them all in the same `MechanicalObject` bothers you?
In the Tripod for instance we had to couple those frames with different articulations, so it was necessary to separate/map them. 
Here if you just want to add a `SlidingActuator` to three of the frames and an `PositionEffector` (I guess) to control the fourth point, they can all live in the same `MechanicalObject`. If that's not what you wanna do just ignore my last comment ^^
